### PR TITLE
Add AIS evaluation for lntm model.

### PR DESCRIPTION
**Pull request by csy530216**
- Add AIS evaluation for lntm
- Support multi-chain inference in lntm
- Change BDMC to AIS since BDMC seems not to work well
- Use a sigmoidal temperature schedule in AIS

### CODE REVIEW

**Comment by thjashin** [comment]
Cool work! THX. Let @cjf00000 have a look first, and I will do the final round before merge.

File: examples/topic_models/lntm_mcem.py [hunk 1]
```diff
@@ -67,42 +73,46 @@ def lntm(observed, D, K, V, eta_mean, eta_logstd):
...
-        # [D], [K], [K]
+        model = lntm(observed, n_chains, D, K, V, eta_mean, eta_logstd)
+
         log_p_eta, log_p_beta = \
             model.local_log_prob(['eta', 'beta'])
 
         theta = tf.nn.softmax(observed['eta'])
         phi = tf.nn.softmax(observed['beta'])
-        pred = tf.matmul(theta, phi)
+        tiled_phi = tf.tile(tf.expand_dims(phi, 0), [n_chains, 1, 1])
```
Comment (cjf00000) [inline examples/topic_models/lntm_mcem.py]: It should be better to reshape theta and multiple with phi, and reshape back. This saves memory.

theta = tf.reshape(theta, [-1, K])
pred = tf.matmul(theta, phi)
pred = tf.reshape(pred, [n_chains, D, -1])

File: examples/topic_models/lntm_mcem.py [hunk 2]
```diff
@@ -115,6 +125,51 @@ def e_obj(observed):
     for i in params:
         print(i.name, i.get_shape())
 
+    # Below is the evaluation part.
+    # Variables whose name starts with '_' is only used in the evaluation part,
+    # to be distinguished from those variables used in the training part above.
+
+    _D = X_test.shape[0]
```
Comment (cjf00000) [inline examples/topic_models/lntm_mcem.py]: We can assume testing batch size = training batch size right now.

Alternatively, we can make D as a placeholder.

File: examples/topic_models/lntm_mcem.py [hunk 3]
```diff
@@ -115,6 +125,51 @@ def e_obj(observed):
...
+    _T = np.sum(X_test)
+    _n_chains = 25
+    _n_temperatures = 1000
+
+    _x = tf.placeholder(tf.float32, shape=[_D, V], name='x')
+    _eta = tf.Variable(tf.zeros([_n_chains, _D, K]), name='eta')
```
Comment (cjf00000) [inline examples/topic_models/lntm_mcem.py]: Reuse train's placeholders and variables

File: examples/topic_models/lntm_mcem.py [hunk 4]
```diff
@@ -115,6 +125,51 @@ def e_obj(observed):
...
+        _theta = tf.nn.softmax(observed['eta'])
+        phi = tf.nn.softmax(observed['beta'])
+        tiled_phi = tf.tile(tf.expand_dims(phi, 0), [_n_chains, 1, 1])
+        _pred = tf.matmul(_theta, tiled_phi)
+
+        tiled_x = tf.tile(tf.expand_dims(observed['x'], 0), [_n_chains, 1, 1])
+
+        _log_px = tf.reduce_sum(tiled_x * tf.log(_pred), -1)
+
+        return _log_prior + _log_px
```
Comment (cjf00000) [inline examples/topic_models/lntm_mcem.py]: Reuse train's joint_obj

File: examples/topic_models/lntm_mcem.py [hunk 5]
```diff
@@ -115,6 +125,51 @@ def e_obj(observed):
...
+    _T = np.sum(X_test)
+    _n_chains = 25
+    _n_temperatures = 1000
+
+    _x = tf.placeholder(tf.float32, shape=[_D, V], name='x')
+    _eta = tf.Variable(tf.zeros([_n_chains, _D, K]), name='eta')
+
+    def _log_prior(observed):
+        _model = lntm(observed, _n_chains, _D, K, V, eta_mean, eta_logstd)
+        return _model.local_log_prob('eta')
```
Comment (cjf00000) [inline examples/topic_models/lntm_mcem.py]: Reuse train's joint_obj

File: examples/topic_models/lntm_mcem.py [hunk 6]
```diff
@@ -167,6 +222,22 @@ def e_obj(observed):
...
+        print('Start evaluation! {} temperatures, {} chains'
+              .format(_n_temperatures, _n_chains))
+        time_ais = -time.time()
+
+        ll_lb = _ais.run(sess, feed_dict={_x: X_test,
+                                          eta_mean: Eta_mean,
+                                          eta_logstd: Eta_logstd})
+
+        time_ais += time.time()
+
```
Comment (cjf00000) [inline examples/topic_models/lntm_mcem.py]: Maybe we need an empty run to initialize the step size?

Or we can make _hmc's step size as a placeholder, and turn off the adaptation. We feed step_size=hmc_info.step_size here.

**Pull request merged**